Move to next year when 12 months are added to first payment date

Mayisha opened the pull request:
Fixes #652 

## Description
When calculating the first renewal date for a synced subscription, we added a year for months less than the current month.
✅ If the current month is August and the month for the first renewal date is in September to December, we do not add a year.
✅ If the current month is August and the month for the first renewal date is in January to July, we add a year.
🔴 If the current month is August and the month for the first renewal date is also in August, we do not add a year. This is the cause of the issue. 

The changes made in this PR will now add a year when the first renewal date is in January to August for the above example. 

## How to test this PR
- Detailed reproduction steps are shared in the issue description https://github.com/Automattic/woocommerce-subscriptions-core/issues/652.
- Confirm that the issue is not reproducible in this branch.
- Confirm that these changes caused no regression with the sync feature. Test this with and without [Prepaid for WooCommerce Subscriptions](https://woocommerce.com/products/prepaid-for-woocommerce-subscriptions/) plugin. [⚠️  IMPORTANT]

Mayisha committed:
move to next year when 12 months are added to first payment date

james-allan commented:
@Mayisha there's currently an issue on this branch if you have a monthly product that is synced to a date later this month. eg if you create a synced product for the 15th (today's date is the 12th). 

If you add that product to your cart, the next payment will be in 1 year + 2 weeks, not in 2 weeks.



| `trunk` | This branch |
|--------|--------|
| <img width="622" alt="Screenshot 2024-08-12 at 10 37 43 AM" src="https://github.com/user-attachments/assets/4bffcdf9-4d86-4376-9214-0653c5dba869"> | <img width="618" alt="Screenshot 2024-08-12 at 10 36 40 AM" src="https://github.com/user-attachments/assets/57526d5b-5dd4-4677-a4c1-cba87dd9de98"> | 


In this case we shouldn't be adding a year.

james-allan commented on includes/class-wc-subscriptions-synchroniser.php:
```diff
@@ -687,7 +687,7 @@ public static function calculate_first_payment_date( $product, $type = 'mysql',
 			}
 			// when a certain number of months are added and the first payment date moves to next year
-			if ( $month_number < gmdate( 'm', $from_timestamp ) ) {
+			if ( $month_number <= gmdate( 'm', $from_timestamp ) ) {
```
I wonder if we need to move this whole if-else block. This whole section seems far more complicated than it needs to be, however it seems like in the if-block above the logic is determining if the customer is purchasing before or after the payment day. 

I might be wrong, however I think we'd only ever need to add a year if the customer is purchasing after the day. ie we're adding a month. 

If that's the case, can we move this if statement up into that case. ie we only need to add a year if we've added a month

james-allan replied:
wait no, that wouldn't be possible in cases when we have a product with x months as the interval. In the issue the product has a 12 month interval and in cases like that we do add x number of months in both cases that could mean we need to add a year.

Mayisha committed:
consider number of months

Mayisha committed:
add changelog entry

Mayisha commented:
Hey @james-allan 

I have made a little change in [fb4d1a2](https://github.com/Automattic/woocommerce-subscriptions-core/pull/651/commits/fb4d1a244346c66fe785cdc336c13105729f36a1) and now adding a year if billing interval is 12 months or more.

Let me know if you can think of any other regression case. Changing anything around synchronization and renewal dates makes me nervous 😬

james-allan commented on includes/class-wc-subscriptions-synchroniser.php:
```diff
@@ -686,8 +686,10 @@ public static function calculate_first_payment_date( $product, $type = 'mysql',
 					$month_number = gmdate( 'm', wcs_add_months( $from_timestamp, $interval ) );
 				}
 			}
+			$number_of_months = WC_Subscriptions_Product::get_interval( $product );
```
I think we might need to use the `$interval` value and not the a fresh value from `WC_Subscriptions_Product::get_interval()`.

The reason is because `$interval` comes with the logic around whether the customer is being charged upfront. ie because the customer is being charged upfront, the customer is entitled to the billing interval in part or in full. If they aren't being charged upfront, they are paying either $0 or a prorated amount and so should be given the next possible payment date available (ie don't add a year).

Mayisha replied:
> The reason is because $interval comes with the logic around whether the customer is being charged upfront. ie because the customer is being charged upfront, the customer is entitled to the billing interval in part or in full. If they aren't being charged upfront, they are paying either $0 or a prorated amount and so should be given the next possible payment date available (ie don't add a year).

Thanks for the explanation @james-allan. This did not come to my mind.
I have addressed this in [d9b4f21](https://github.com/Automattic/woocommerce-subscriptions-core/pull/651/commits/d9b4f21a5515f22a51fec63709ab275008572c54)

james-allan reviewed (changes requested):
> Let me know if you can think of any other regression case. Changing anything around synchronization and renewal dates makes me nervous 😬

me too 😕 interpreting this code is so difficult.

There's a couple of scenarios to test given the changes here are limited. It's synced subscriptions with a monthly billing period. So I tested the following scenarios:

**PRORATED FIRST PAYMENT**
_In this case we expect the customer to pay $x upfront and get a next available payment date_
 
| Every (interval) | Synced to | is Before or After today? | Payment date result | 
|:--------:|:--------:|:--------:|:--------:|
| 1 Month | 1st | Before | September 1, 2024 👍  |
| 1 Month | 20th | After | August 20, 2024 👍  |
| 6 Months | 1st | Before | September 1, 2024 👍 |
| 6 Months | 20th | After | August 20, 2024 👍 |
| 1 Month 2 week trial | 1st | Before | September 1, 2024 👍 |
| 1 Month 2 week trial | 20th | After | September 20, 2024 👍 (because the trial ends after the 20th of Aug, it's pushed to next month) |
| 12 Months * | 1st | Before | August 1, 2025 ❌ |
| 12 Months * | 20th | After | August 20, 2025 ❌ |

\* to set a 12 month subscription interval you need custom code or to set it in the DB. 

The problem in the 12 month case is caused by the or condition you've added in your most recent changes. 

---

**FULL UPFRONT PAYMENT**
_This case is super weird because I don't actually know what the intended behaviour is. In some cases it calculates a first date after 1 billing period and sometimes it doesn't._
 
| Every (interval) | Synced to | is Before or After today? | Payment date result | 
|:--------:|:--------:|:--------:|:--------:|
| 1 Month | 1st | Before | September 1, 2024 (in ~2 weeks) |
| 1 Month | 20th | After | August 20, 2024 (in 5 days) |
| 6 Months | 1st | Before | February 1st, 2025 (in 5 months and 2 weeks) |
| 6 Months | 20th | After | January 20, 2025 (in 5 months and 5 days) |
| 1 Month 2 week trial | 1st | Before | September 1, 2024 |
| 1 Month 2 week trial | 20th | After | September 20, 2024 |
| 12 Months * | 1st | Before | July 1, 2025 ⁉️ |
| 12 Months * | 20th | After | August 20, 2025 👍 |

⁉️ Reading the documentation around this "upfront cost" sync setting ([doc](https://woocommerce.com/document/subscriptions/renewal-synchronisation/#prorate-settings)), it's intended to be used for stores that sell a subscription box and so getting "this months box" by paying upfront. I think that make sense and means that the customer is entitled to some or all of the billing period is probably fine. Still, the inconsistency is weird. It's unrelated to this PR anyway. 

All of this testing to say, the sync calculations when paying upfront are weird, but I think your changes look close, we just need to fix the situation where the customer pays nothing or a prorated amount upfront. I left an inline comment about that.

Mayisha committed:
use value of $interval variable instead of calling 'get_interval'

Mayisha commented:
I have addressed the feedback in [d9b4f21](https://github.com/Automattic/woocommerce-subscriptions-core/pull/651/commits/d9b4f21a5515f22a51fec63709ab275008572c54) and using the value of $interval to compare. It seems to fix the dates for `PRORATED FIRST PAYMENT` mentioned [here](https://github.com/Automattic/woocommerce-subscriptions-core/pull/651#pullrequestreview-2239808275) 🎉 

However, I am still a bit confused with the `FULL UPFRONT PAYMENT` cases. I see the dates different than what you have noted (I have checked both before and after the last commit)

**With sync day after today (22nd of each month) I get July 22, 2025**
<img width="486" alt="Screenshot 2024-08-20 at 2 40 03 AM" src="https://github.com/user-attachments/assets/79a6284e-cc7b-4c62-85c1-0606f38bfcf7">
 
**With sync day before today (1st of each month) I get August 01, 2025**
<img width="486" alt="Screenshot 2024-08-20 at 2 30 12 AM" src="https://github.com/user-attachments/assets/5c5d2aba-ed3c-4bc3-9d3f-77d11f233470">

Mayisha committed:
Merge branch 'trunk' into fix/synced-prepaid-sub

Mayisha committed:
update version number in changelog

james-allan commented:
> However, I am still a bit confused with the FULL UPFRONT PAYMENT cases. I see the dates different than what you have noted (I have checked both before and after the last commit)

I think those dates you're getting are correct and in my tests today, they are the same as what I'm getting. 

| After | Before |
|--------|--------|
| <p align="center"><img width="634" alt="Screenshot 2024-08-20 at 5 00 41 PM" src="https://github.com/user-attachments/assets/f0ee25c7-17d7-49d5-b436-2eb578dc060b"></br><sup><strong>22 July 2025</strong> when purchasing subscription synced <strong>after</strong> today.</sup></p>| <p align="center"><img width="603" alt="Screenshot 2024-08-20 at 5 03 06 PM" src="https://github.com/user-attachments/assets/d8ed9692-5bbf-4474-b4ed-1127a2ec4516"></br><sup><strong>2 Aug 2025</strong> when purchasing subscription synced <strong>before</strong> today.</sup></p> | 

In my table of results I had it around the other way -- July 2025 when purchasing **before** and Aug 2025 when purchasing **after**. That was wrong. The reason July is returned is because of the [`- 1 month` in the `$from_day <= $payment_day` case](https://github.com/Automattic/woocommerce-subscriptions-core/blob/7.4.0/includes/class-wc-subscriptions-synchroniser.php#L675-L676) so it couldn't have ever returned a July date in the **"before"** case. Before and after are getting a little confusing because the code refers to before as `$from_date` before.

Anyway, that's just a very confusing way of saying that I think this is correct now. There's still the strangeness around why July is returned at all, but that's not related to this PR.
